spec/virtual_includes_spec.rb

```diff
@@ -228,6 +240,78 @@
     end
   end
 
+  context ".merge_includes" do
+    it "merges when first is blank" do
+      first  = {}
+      second = {:key => {}}
+      result = {:key => {}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges when second is blank" do
+      first  = {:key => {}}
+      second = {}
+      result = {:key => {}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "keeps other keys" do
+      first  = {:other1 => {}}
+      second = {:other2 => {}}
+      result = {:other1 => {}, :other2 => {}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges blank, blank" do
+      first  = {:key => {}}
+      second = {:key => {}}
+      result = {:key => {}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges blank, hash" do
+      first  = {:key => {}}
+      second = {:key => {:more2 => {}}}
+      result = {:key => {:more2 => {}}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges hash, blank" do
+      first  = {:key => {:more1 => {}}}
+      second = {:key => {}}
+      result = {:key => {:more1 => {}}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges hash, hash - 1 level only" do
+      first  = {:key => {:more1 => {}}}
+      second = {:key => {:more2 => {}}}
+      result = {:key => {:more1 => {}, :more2 => {}}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges hash(hash), hash(symbol)" do
+      first  = {:key => {:more1 => {}}}
+      second = {:key => :more2}
+      result = {:key => {:more1 => {}, :more2 => {}}}
+      expect(Author.merge_includes(first, second)).to eq(result)
+    end
+
+    it "merges hash(hash), hash(array)" do
+      first  = {:key => {:more1 => {}}}
+      second = {:key => [:more2]}
+      result = {:key => {:more1 => {}, :more2 => {}}}
```

Comment:
Head asplode

Reply:
It feels like that's not going to be obvious... do you have an example where includes look like this?

Reply:
[Host] has a bunch of different `:uses`:

```ruby
class Host
  virtual_column :v_owning_datacenter, :type => :string, :uses => :all_relationships
  virtual_column :os_image_name,       :type => :string, :uses => [:operating_system, :hardware]
  virtual_has_many :miq_scsi_luns, :uses => {:hardware => {:storage_adapters => {:miq_scsi_targets => :miq_scsi_luns}}}
end
```

It has a key, an array, and a hash pointing to a hash pointing to a hash pointing to a key.

---

[Flavor] shows a few styles of `includes()`

```ruby
class Flavor
  def self.tenant_joins_clause(scope)
    scope.includes(:cloud_tenants => "source_tenant", :ext_management_system => {})
         .references(:cloud_tenants => "source_tenant", :ext_management_system => {})
  end
end
```

It has a hash key pointing to a string and a hash key pointing to a hash.

[Host]: https://github.com/ManageIQ/manageiq/blob/master/app/models/host.rb#L118
[Flavor]: https://github.com/ManageIQ/manageiq/blob/master/app/models/flavor.rb#L46

Reply:
and what does it do if there are multiple values in the array:

this?
```
      first  = {:key => {:more1 => {}}}
      second = {:key => [:more2, :more3]}
      result = {:key => {:more1 => {}, :more2 => {}, :more3 => {}}}
```

Reply:
yup